tiatoolbox/wsicore/wsireader.py

```diff
@@ -2389,6 +2314,207 @@ def read_bounds(
+                pad_mode=pad_mode,
+                pad_constant_values=pad_constant_values,
+            )
+
+        # Find parameters for optimal read
+        (
+            read_level,
+            level_location,
+            level_read_size,
+            post_read_scale,
+            _,
+        ) = self.find_read_rect_params(
+            location=location,
+            size=size,
+            resolution=resolution,
+            units=units,
+        )
+
+        wsi = self.wsi
+
+        # Read at optimal level and corrected read size
+        level_size = self.info.level_dimensions[read_level]
+        constrained_read_bounds = utils.image.find_overlap(
+            read_location=level_location,
+            read_size=level_read_size,
+            image_size=level_size,
+        )
+        _, constrained_read_size = utils.transforms.bounds2locsize(
+            constrained_read_bounds
+        )
+        im_region = wsi.read_region(location, read_level, constrained_read_size)
+        im_region = np.array(im_region)
+
+        # Apply padding outside of the slide area
+        level_read_bounds = utils.transforms.locsize2bounds(
+            level_location, level_read_size
+        )
+        im_region = utils.image.crop_and_pad_edges(
+            bounds=level_read_bounds,
+            max_dimensions=level_size,
+            region=im_region,
+            pad_mode=pad_mode,
+            pad_constant_values=pad_constant_values,
+        )
+
+        # Resize to correct scale if required
+        im_region = utils.transforms.imresize(
+            img=im_region,
+            scale_factor=post_read_scale,
+            output_size=tuple(np.array(size).astype(int)),
+            interpolation=interpolation,
+        )
+
+        return utils.transforms.background_composite(image=im_region)
+
+    def read_bounds(
+        self,
+        bounds,
+        resolution=0,
+        units="level",
+        interpolation="optimise",
+        pad_mode="constant",
+        pad_constant_values=0,
+        coord_space="baseline",
+        **kwargs,
+    ):
+        # convert from requested to `baseline`
+        bounds_at_baseline = bounds
+        if coord_space == "resolution":
+            bounds_at_baseline = self._bounds_at_resolution_to_baseline(
+                bounds, resolution, units
+            )
+            _, size_at_requested = utils.transforms.bounds2locsize(bounds)
+            # don't use the `output_size` (`size_at_requested`) here
+            # because the rounding error at `bounds_at_baseline` leads to
+            # different `size_at_requested` (keeping same read resolution
+            # but base image is of different scale)
+            (
+                read_level,
+                bounds_at_read_level,
+                _,
+                post_read_scale,
+            ) = self._find_read_bounds_params(
+                bounds_at_baseline, resolution=resolution, units=units
+            )
```

Comment:
At some point we should put it into baseclass

Reply:
Can make a note. Probably should be another PR to avoid doing too much in one.